### PR TITLE
Link to Study Dataset Category Assignment

### PR DESCRIPTION
**Pull request by RosalineP**
#### Rationale
It is desired that we add functionality to be able to categorize the dataset resulting from a link to study at the assay design level, for the auto-link case. For the manual link to study case, the dataset category assignment can now happen at the initial 'publishStart' page.

#### Related Pull Requests
* https://github.com/LabKey/labkey-ui-components/pull/604

#### Changes
* Roundtrip new 'autoLinkCategory' property
* Support dataset category assignment for auto-link, manual link, and manual link using pipeline
* Prevent entering Dataset Category name in Assay Design that's longer than 300 characters
* Add new 'Specify Linked Dataset Category' text input on 'publishStart' page

### CODE REVIEW

File: study/src/org/labkey/study/assay/StudyPublishManager.java [hunk 1]
```diff
@@ -1080,7 +1110,16 @@ public ActionURL autoLinkAssayResults(ExpProtocol protocol, ExpRun run, User use
...
                 final Container targetStudyContainer = ContainerManager.getForId(targetStudyContainerId);
 
-                return autoLinkResults(protocol, provider, run, user, container, targetStudyContainer, errors, LOG);
+                // Determine if the category is predefined
+                ViewCategory targetStudyCategory = null;
+                if (protocol.getObjectProperties().get(StudyPublishService.AUTO_LINK_CONTAINER_PROPERTY_URI) != null)
+                {
+                    String categoryName = protocol.getObjectProperties().get(StudyPublishService.AUTO_LINK_CONTAINER_PROPERTY_URI).getStringValue();
+                    targetStudyCategory = ViewCategoryManager.getInstance().ensureViewCategory(targetStudyContainer, user, categoryName);
+                    LOG.debug("Determined predefined Dataset Category to assign, " + targetStudyCategory.getLabel());
```
Comment (RosalineP) [inline study/src/org/labkey/study/assay/StudyPublishManager.java]: Considering pushing down the category name string instead, and doing the `ensureViewCategory` closer to the `_publishData` layer—thoughts?

File: api/src/org/labkey/api/study/publish/publishChooseStudy.jsp [hunk 2]
```diff
@@ -213,6 +214,20 @@
         }
     %>
 
+    <%
+        if (autoLinkEnabled)
+        {
+    %>
+    <div class="form-group">
+        <label for="autoLinkCategory" class="control-label col-sm-3 col-lg-2"> Specify Linked Dataset Category </label>
```
Comment (RosalineP) [inline api/src/org/labkey/api/study/publish/publishChooseStudy.jsp]: Considering adding a helptext, pending convo with Molly.

Comment (RosalineP) [reply api/src/org/labkey/api/study/publish/publishChooseStudy.jsp]: Resolved

File: study/src/org/labkey/study/assay/StudyPublishManager.java (hunk 1 above)
Comment (labkey-klum) [reply study/src/org/labkey/study/assay/StudyPublishManager.java]: That might be good, especially if you can push it into the transaction so if the publish fails we don't create an unused category in the process.

File: api/src/org/labkey/api/study/publish/StudyPublishService.java [hunk 3]
```diff
@@ -62,6 +62,7 @@
     String TARGET_STUDY_PROPERTY_NAME = "TargetStudy";
 
     String AUTO_LINK_TARGET_PROPERTY_URI = "terms.labkey.org#AutoCopyTargetContainer";
+    String AUTO_LINK_CONTAINER_PROPERTY_URI = "terms.labkey.org#AutoLinkCategory";
```
Comment (labkey-klum) [inline api/src/org/labkey/api/study/publish/StudyPublishService.java]: should this be category instead of container?

**Review by labkey-klum — APPROVED** [review]

File: api/src/org/labkey/api/study/publish/StudyPublishService.java (hunk 3 above)
Comment (RosalineP) [reply api/src/org/labkey/api/study/publish/StudyPublishService.java]: Oop, yes. Resolved!

File: study/src/org/labkey/study/assay/StudyPublishManager.java (hunk 1 above)
Comment (RosalineP) [reply study/src/org/labkey/study/assay/StudyPublishManager.java]: Done!

**Pull request merged**